.github/workflows/main.yaml

```diff
@@ -0,0 +1,170 @@
+      - name: Push images
+        if: ${{ (steps.changed-files.outputs.any_changed == 'true') }}
+        run: docker push ${{ steps.meta.outputs.tags }}
+      - name: Set Docker Tag
+        if: ${{ (steps.changed-files.outputs.any_changed == 'false') && (github.event_name == 'pull_request') }}
+        run: echo "DOCKER_TAG=ghcr.io/${{github.repository_owner}}/cobalt-${docker_service}:${GITHUB_BASE_REF}" >> $GITHUB_ENV
+      - name: Set Docker Tag
+        if: ${{ (steps.changed-files.outputs.any_changed == 'false') && (github.event_name != 'pull_request') }}
+        run: echo "DOCKER_TAG=ghcr.io/${{github.repository_owner}}/cobalt-${docker_service}:${GITHUB_REF_NAME}" >> $GITHUB_ENV
+      - name: Show Docker Tag
+        run: echo $DOCKER_TAG
+    outputs:
+      docker_tag: ${{ env.DOCKER_TAG }}
+
+  # Generates build matrix based on json configuration file.
+  build-matrix:
+    runs-on: ubuntu-latest
+    steps:
+      - id: checkout
+        uses: actions/checkout@v3
+      - id: set-platforms
+        run: echo "::set-output name=platforms::$(cat ${GITHUB_WORKSPACE}/.github/config/${{ inputs.platform }}.json | jq -c '.platforms')"
+      - id: set-includes
+        run: echo "::set-output name=includes::$(cat ${GITHUB_WORKSPACE}/.github/config/${{ inputs.platform }}.json | jq -c '.includes')"
+    outputs:
+      platforms: ${{ steps.set-platforms.outputs.platforms }}
+      includes: ${{ steps.set-includes.outputs.includes }}
+  # Runs Cobalt CI workflow.
+  build:
+    needs: [docker-image, build-matrix]
+    runs-on: [self-hosted, linux, X64]
+    name: ${{matrix.platform}}_${{matrix.config}}
+    strategy:
+      fail-fast: false
+      matrix:
+        platform: ${{ fromJson(needs.build-matrix.outputs.platforms) }}
+        include: ${{ fromJson(needs.build-matrix.outputs.includes) }}
+        config: [devel, debug, qa, gold]
+    container: ${{ needs.docker-image.outputs.docker_tag }}
+    steps:
+      - name: checkout
+        uses: actions/checkout@v2
+      - name: Set env vars
+        run: |
+          echo "PYTHONPATH=$GITHUB_WORKSPACE" >> $GITHUB_ENV
+          echo "STARBOARD_TOOLCHAINS_DIR=/root/starboard-toolchains" >> $GITHUB_ENV
+      - name: Set Android env vars
+        if: startsWith(${{matrix.target_platform}}, 'android')
+        run: |
+          echo "ANDROID_HOME=/root/starboard-toolchains/AndroidSdk/" >> $GITHUB_ENV
+          echo "GCS_NIGHTLY_PATH=gs://cobalt-actions-devel-build-artifacts" >> $GITHUB_ENV
+      - name: Print env vars
+        run: env | sort
+      - name: "gn gen"
+        run: |
+          gn gen $GITHUB_WORKSPACE/out/${{matrix.target_platform}}_${{matrix.config}} --args="target_platform=\"${{matrix.target_platform}}\" ${{matrix.use_asan}} ${{matrix.is_clang}} ${{matrix.using_old_compiler}} ${{matrix.sb_api_version}} ${{matrix.target_os}} ${{matrix.target_cpu}} is_internal_build=false build_type=\"${{matrix.config}}\""
```

Comment:
Instead of referencing each possible gn arg here, we should let build configurations use a field for arbitrary gn arguments, like `"extra_gn_arguments": "use_asan=false using_old_compiler=true"`
Maybe we can have an exception for `sb_api_version`, but for most I think they should just go into an "other" bucket.

Reply:
Ack.